### DISCUSSION TITLE
Large environment data

### DISCUSSION THREAD

**Discussion by isolin** [FLAMEGPU/FLAMEGPU2#793] [Q&A] [answered]
On my journey of building an ecosystem simulation (see [here](https://github.com/FLAMEGPU/FLAMEGPU2/discussions/778#discussioncomment-2143944)) I encountered the following problem. I would like to run the simulation in time steps of a few minutes over the course of several years. E.g. one year in 10 minute steps takes `52560` simulation steps. I am using the python interface to take advantage of various libraries that help me to precompute environmental data, like precipitation amount for each time step.

For the example above it makes 52560 * 4 bytes = that is just 205 kB

That is not much at all. 4K textures are way larger. I hoped for storing up to 100x more, but already for 52560*4 bytes I am getting a `FLAMEGPURuntimeException: (OutOfMemory) /__w/FLAMEGPU2/FLAMEGPU2/src/flamegpu/runtime/utility/EnvironmentManager.cu(86): Insufficient EnvProperty memory to create new properties,in EnvironmentManager::init().`

How can I pass larger data to the kernels?

Update: I see that macro environment properties might be the way to go, but the data I need to store is constant. However, I didn't find any example how to populate them with data before the simulation starts. The only examples show creating an empty array like `env.newMacroPropertyInt("mi_prop", 3, 3, 3, 3);  # Create int array macro property 'mi_prop', with dimensions of [3, 3, 3, 3]`.

**Comment by Robadob** [comment, marked as answer]
You are correct, that macro properties are intended for large environment properties. Regular environment properties are stored in CUDA constant memory which if I recall correctly is limited to 32KB, of which I think we reserve about half for environmental properties.

It's not currently possible to give macro properties values at model description time, I would suggest you initialise them with an init function. You can access external python variables within initialisation functions to fill them however you like (although I believe you will need to loop through setting each element individually due to how they are implemented).

[](https://docs.flamegpu.com/guide/3-behaviour-definition/6-host-functions.html)

**Reply by Robadob** [reply]
Alternatively, you could store your precipitation data on the host, and simply update a small environment property each time step with the current precipitation data using a host layer/step function.

**Reply by ptheywood** [reply]
The CUDA constant memory space is 64kB in size for all current CUDA architectures ([source](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#features-and-technical-specifications__technical-specifications-per-compute-capability))

An alternative would be to use a single value in an Environment Property which is updated from a step function every iteration, or an environment property of N elements, updated every N steps (with an extra value required to track the position within the window). I'm not sure which option would offer the best runtime performance.

**Reply by isolin** [reply]
@ptheywood My original approach was having a single variable that kind-of randomly changed every step. But that was way too random, I need to follow some global constraints (e.g. the sum of values over a given time frame and a specific time-dependent probability distribution). So it was the easiest for me to grab numpy and skyfield to precompute all I need, without trying to reinvent the wheel inside of the kernels, hoping that my 12GB VRAM should be able to occupy a few MB of precomputed data. I definitely want it precomputed.

**Reply by isolin** [reply]
@Robadob could you please elaborate more on storing the data on the host and fetching it in each step? Is there please any example or snippet showing what you mean? Or anything related to the following would help as well:
> You can access external python variables within initialisation functions to fill them however you

I think I found it in the [docs](https://docs.flamegpu.com/guide/3-behaviour-definition/6-host-functions.html). Let's see if I can make it work.

**Reply by Robadob** [reply]
I don't think there's an official example, you would want something like this

```py
# Precompute array of precipitation data
my_precipitation_data = np.zeros(1024)
for i in my_precipitation_data:
  i = np.random()



# Define an host function called update_precipitation
class update_precipitation(pyflamegpu.HostFunctionCallback):
  def run(self, FLAMEGPU):
    # Retrieve the current step index
    step_no = FLAMEGPU.getStepCounter()
    # Retrieve the current precipitation value
    current_precipitation = my_precipitation_data[step_no % len(my_precipitation_data)]
    # Update precipitation environment property
    FLAMEGPU.environment.setPropertyFloat("precipitation", current_precipitation)

    
# Then as part of model description
model.addStepFunctionCallback(update_precipitation().__disown__());
# Or add it to a specific layer where you would like it called
model.newLayer().addHostFunctionCallback(update_precipitation().__disown__());

```

FGPU2 host functions in the Python API are implemented as proper Python classes, so you could even store the precipitation data inside the class.

```py
# Define an host function called update_precipitation
class update_precipitation(pyflamegpu.HostFunctionCallback):
  def __init__(self):
    super().__init__()
    self.my_precipitation_data = np.zeros(1024)
    for i in my_precipitation_data:
      i = np.random()

  def run(self, FLAMEGPU):
    # Retrieve the current step index
    step_no = FLAMEGPU.getStepCounter()
    # Retrieve the current precipitation value
    current_precipitation = self.my_precipitation_data[step_no % len(self.my_precipitation_data)]
    # Update precipitation environment property
    FLAMEGPU.environment.setPropertyFloat("precipitation", current_precipitation)
```

**Reply by isolin** [reply]
Awesome, works like a charm! Thank you very much for the help!
